Approved/remove-new-array.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Remove default-initialized new array expressions
+description: The `new Type[N]` array expression is deprecated and removed from Q#.
+author: Sarah Marshall
+date: October 28, 2020
+---
+
+# Proposal
+
+The `new Type[N]` expression, which creates an array of `N` values using the default value of `Type`, is deprecated and removed from Q#.
+
+# Justification
+
+The existence of `new Type[N]` has negative effects on the language.
+
+The most important negative effect is the assumption that every type has a default value.
+This is not a reasonable assumption, because it is not possible to define useful default values of types like `Qubit` and `a -> b`.
+Their current default values are invalid and will trigger an error if they are used.
+But the syntax is the same both for types with valid default values, and those without, which is misleading: while `new Int[3]` is a well-defined value whose items can be used safely, `new Qubit[3]` creates a potential trap that will crash the program if any item in the array is used before being set to a valid value first.
+
+This assumption is also invalid in the presence of uninhabited types such as `Void`.
+If Q# requires that every type has a default value, then uninhabited types are not possible to express properly.
+This can even cause subtle bugs in the soundness of a type system, as demonstrated by [this bug in Java's generics](https://hackernoon.com/java-is-unsound-28c84cb2b3f) that happened because of the existence of a value (specifically `null`) for a type that should have been uninhabited.
+
+Finally, the use of the `new` keyword is inconsistent with other ways to create values in Q#.
+`new` is required only when creating a default-initialized array, not when using a square bracket array literal or when creating values of other types.
+
+# Description
+
+## Current Status
+
+Q# currently supports this syntax:
+
+```qsharp
+// Create an empty array of type Int[].
+let empty = new Int[0];
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = new Bool[10];
+```
+
+## Proposed Modification
+
+The syntax above will be removed.
+The same functionality can be accomplished with functions in the standard library:
+
+```qsharp
+// Create an empty arra of type Int[].
+let empty = EmptyArray<Int>();
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = ConstantArray(10, false);
+```
```

Comment:
I agree that removing the `new` keyword is goodness. However, the proposed syntax may feel foreign for developers coming from Python, C#, Java, or C++.  It will also look inconsistent with `using (a = Qubit[10]) {}`. We should explore syntax to create empty and constant arrays that is consistent with both `a = [1, 2, 3];` and the qubit example.  For example, Python has an elegant syntax for these scenarios:
```
a = []
b = [1] * 10
```

Reply:
> However, the proposed syntax may feel foreign for developers coming from Python, C#, Java, or C++.

To clarify, this proposal does not contain any new syntax. `EmptyArray` and `ConstantArray` are functions that are already defined in the Q# standard library. They have counterparts in C# too:

```csharp
var a = Array.Empty<int>();
var b = Enumerable.Repeat(1, 10).ToArray();
```

> ```
> a = []
> ```

Yes, this is my favorite way to create an empty array, but it's blocked for the moment because Q#'s type inference is not sufficient to support this. See microsoft/qsharp-compiler#617.

> ```
> b = [1] * 10
> ```

Because Q# is statically typed, unlike Python, this syntax has some additional implications for us:

* The type of `*` is overloaded to both `('a[], Int) -> 'a[]` and `(Int, 'a[]) -> 'a[]`, since Python also supports `10 * [1]`. Q# does not support overloading. I would argue that there are good reasons for Q# to never support overloading, but there is a feature request for this: microsoft/qsharp-compiler#29
* We would need to choose a different operator name than `*`. Otherwise, it's not possible to reconcile the type of `*` used for arrays with the type of `*` needed to multiply two numbers together. This would make it very difficult to create a type class to represent numeric types. See microsoft/qsharp-compiler#557.

Because of these issues, and because in my opinion, `[1] * 10` is harder to read than `ConstantArray(10, 1)`, I would recommend keeping `ConstantArray` as a normal function instead of adding new syntax to Q#.

Reply:
Having these library functions is fine. I'm concerned with making them the only way to create certain kind of arrays. I don't see how Q# developers will be excited to go through their code to replace `a=new Int[0]` with `a = EmptyArray<Int>()`. It is verbose and not common in other languages. In contrast, replacing `a=new Int[0]` with `a = []` feels like a great change. I think that we should hold on deprecating the `new` keyword until microsoft/qsharp-compiler#617 is fixed. Requiring to replace `a = new Int[1]` with `a = ConstantArray(1, 0)` isn't great either.  We don't have to use the python syntax as is but rather explore similar ideas as alternative options to ConstantArray.

Reply:
> I think that we should hold on deprecating the `new` keyword until microsoft/qsharp-compiler#617 is fixed.

While I agree that `[]` is the best syntax, and that we should try to avoid deprecating syntax in favor of alternatives that will themselves be replaced, we also need to balance the benefits of removing default value semantics from Q# with the effort needed to fix microsoft/qsharp-compiler#617.

1. I think it's important to remove default values by Q# 1.0. If they exist in 1.0, they'll be harder to remove later. By contrast, it's much easier to add `[]` and/or a built-in constant array syntax later on.
2. The effort needed to fully resolve microsoft/qsharp-compiler#617 is large. We could fix some limited cases more easily, but we'd still need some way to explicitly specify the type of an empty array for the remaining cases. There's alternative syntax we could consider instead of `EmptyArray<T>()`, such as `[] : T[]`.
3. The alternative that we pick for `new T[0]` should persist even after `[]` is added - I don't expect `EmptyArray` or `ConstantArray` to ever be deprecated, so while we'll be providing a better alternative eventually, there won't be another round of deprecation like there was with `new`.

> It is verbose and not common in other languages.

For comparison, `Array.Empty<T>()` is preferred over `new T[0]` to create empty arrays in C#: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825 (albeit for different reasons than Q#).

> Requiring to replace `a = new Int[1]` with `a = ConstantArray(1, 0)` isn't great either. We don't have to use the python syntax as is but rather explore similar ideas as alternative options to ConstantArray.

I would be interested in concrete syntax suggestions. I cannot think of any, and other than Python, I am not aware of any language that provides this functionality using operator syntax instead of a named function. I also think this is subjective, and not a clear-cut case where special syntax is definitely needed. The function seems preferable to me, other than this specific case of an `Int` array where it's not immediately clear which parameter is the length and which is the value.

Reply:
> > I think that we should hold on deprecating the `new` keyword until [microsoft/qsharp-compiler#617](https://github.com/microsoft/qsharp-compiler/issues/617) is fixed.
> 
> While I agree that `[]` is the best syntax, and that we should try to avoid deprecating syntax in favor of alternatives that will themselves be replaced, we also need to balance the benefits of removing default value semantics from Q# with the effort needed to fix [microsoft/qsharp-compiler#617](https://github.com/microsoft/qsharp-compiler/issues/617).
> 
> 1. I think it's important to remove default values by Q# 1.0. If they exist in 1.0, they'll be harder to remove later. By contrast, it's much easier to add `[]` and/or a built-in constant array syntax later on.

Absolutely agreed, default values have been a pretty significant thorn in the side of Q#'s type safety for a while now. The sheer breakage one can cause with `H((new Qubit[1])[0]);` is pretty impressive, for instance.

> 2. The effort needed to fully resolve [microsoft/qsharp-compiler#617](https://github.com/microsoft/qsharp-compiler/issues/617) is large. We could fix some limited cases more easily, but we'd still need some way to explicitly specify the type of an empty array for the remaining cases. There's alternative syntax we could consider instead of `EmptyArray<Int>()`, such as `[] : Int[]`.
> 3. The alternative that we pick for `new T[0]` should persist even after `[]` is added - I don't expect `EmptyArray` or `ConstantArray` to ever be deprecated, so while we'll be providing a better alternative eventually, there won't be another round of deprecation like there was with `new`.
> 
> > It is verbose and not common in other languages.
> 
> For comparison, `Array.Empty<T>()` is preferred over `new T[0]` to create empty arrays in C#: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825 (albeit for different reasons than Q#).
> 
> > Requiring to replace `a = new Int[1]` with `a = ConstantArray(1, 0)` isn't great either. We don't have to use the python syntax as is but rather explore similar ideas as alternative options to ConstantArray.

By way of understanding, what isn't great about it? The `new`-based syntax doesn't make clear what value elements are populated with; the user has to read the table of default values for each type in the Q# language docs to know that `new Int[1]` and `[0]` are equivalent. By contrast, `ConstantArray(3, PauliX)` clearly documents to the _reader_ that the length of the new array is 3, and that each element is `PauliX`. In general, that goes well with the focus that Q# should be easy to read, even when that makes _writing_ Q# slightly more cumbersome; especially given difficulty posed by the problem domain of quantum computing, it's important to make sure that code is as readable and understandable as it can reasonably be.

> I would be interested in concrete syntax suggestions. I cannot think of any, and other than Python, I am not aware of any language that provides this functionality using operator syntax instead of a named function. I also think this is subjective, and not a clear-cut case where special syntax is definitely needed.

As a quick clarification, `[value] * length` in Python is an effect of `+` being used for list concatenation. That is, if one follows common intuitions about algebraic expressions, I don't think that `[a, b] + [a, b]` being the same thing as `[a, b] * 2` is wholly unreasonable  —  indeed, the syntax reflects that Python lists are monoids under `+`. That is, Python doesn't have special syntax for "create a list with 𝑛 copies of the same value" so much as it has a general syntax for "repeat this list 𝑛 times." From that perspective, if we want to borrow Python's `[a] * n` syntax, I would strongly suggest we not treat it as a special case, but that we also treat the monoid concept more generally.

> The function seems preferable to me, other than this specific case of an `Int` array where it's not immediately clear which parameter is the length and which is the value.

We could possibly address that by a named input feature, something like `ConstantArray(length = 10, value = 3)`. Happy to think a bit on that and draft a suggestion if that would be helpful.

Reply:
> From that perspective, if we want to borrow Python's [a] * n syntax, I would strongly suggest we not treat it as a special case, but that we also treat the monoid concept more generally.

I'm fine with Python supporting repeating a list multiple times (i.e. repeated application of the monoid operation on lists), my issue is with overloading `*`. :) In general, multiplication is not repeated addition. More concretely it conflicts with the usual type of `* : a -> a -> a`.

Reply:
> > From that perspective, if we want to borrow Python's [a] * n syntax, I would strongly suggest we not treat it as a special case, but that we also treat the monoid concept more generally.
> 
> I'm fine with Python supporting repeating a list multiple times (i.e. repeated application of the monoid operation on lists), my issue is with overloading `*`. :) In general, multiplication is not repeated addition. More concretely it conflicts with the usual type of `(*) : a -> a -> a`.

Fair enough, though that's an abuse perpetuated by group theorists for many years; using `*` to denote repetitions of a monoid addition is an abuse that is not unique to Python. I agree, though, the historical existence of an abuse of notation isn't a great argument for perpetuating it, heh.

Reply:
Sorry for coming late to this thread; good discussions!
"I would argue that there are good reasons for Q# to never support overloading" we already do support some overloading, like + that can be used for all arithmetic types and also for concatenations. 
"We would need to choose a different operator name than *." Yes, I agree that this would indeed be good in this case.

I think we all want to support empty array literals (`[]`), and the only question is around priority here. The priority conversation is a separate one to have outside this proposal. What we do need to decide for this proposal here is whether or not having support for empty array literals is a blocker for removing new-array expressions. I don't think it is, but I am happy to chat about whether it should be. 
I wouldn't introduce something like `[] : Int[]` as a "one-off" thing for new arrays only. If we were to introduce something like that syntax, then is should probably be supported for all expressions in general. An alternative to that would be to allow explicit type annotations for local bindings (`let a : Int[] = [];`). Both of these require careful consideration and are more work to implement than supporting `[]` without any annotation. 

"if we want to borrow Python's [a] * n syntax, I would strongly suggest we not treat it as a special case, but that we also treat the monoid concept more generally." Yes, fully agreed.

Reply:
> We already do support some overloading, like + that can be used for all arithmetic types and also for concatenations.

The operators in Q# right now sort of exist outside of the type system, so I don't think it's overloading exactly. The issue I am thinking of is if/when we want to define `*` as a function in Q#, we need to give it a proper type. That's not possible without overloading in the cases of numeric multiplication and list multiplication.

For `+` specifically, depending on how we define it, it doesn't necessarily require overloading. It at least has the same type in both cases: lists are `('a[], 'a[]) -> 'a[]`, which is still an instance of `('a, 'a) -> 'a` that you would expect for numeric `+`. But we may want to make it part of a different type class (Haskell separates them), in which case we would need to either give it a separate name or introduce overloading into the type system proper.

Reply:
> I wouldn't introduce something like `[] : Int[]` as a "one-off" thing for new arrays only.

I agree, and I wouldn't propose that we do that. If we used `[] : Int[]`, I would agree that it should be type annotations for all expressions.

Reply:
Let me summarize some of our conversation today. 

Looking in more detail at what operators make sense for matrix multiplication, we think that there is not much need beyond matrix multiplication, broadcasting, and maybe an operator for a tensor product (that can be covered by the broadcasting, but is pretty unreadable that way). Python additionally supports concatenation for multidimensional arrays, but is is unclear whether there would be much value for us to do that same. Out of these, there is no syntax that naturally covers the scenario of array construction. 

We have had an extensive conversation about the value of modules. The clear upside to using modules over free floating callables, leading to an Array.Constant syntax over ConstantArray, is the additional discoverability of callables that can be easily identified via an appropriate module. For callables that should exist not just for arrays but more broadly for collections/enumerables in the future, such as e.g. Mapped, it makes more sense to have them not be part of an Array module since it would imply that we would then in the future have not just Array.Mapped but also List.Mapped, instead of simply a parameterized version of Mapped. In the case of Constant, the argument alone does not make it clear whether an array or e.g. a list should be constructed. However, with the same effort that is needed to support empty array literals `[]`, it is possible to infer the type argument for the return type such that the same considerations as for Mapped are also applicable to Constant. 

Replacing `new Int[10]` simply with `Constant(10)` or e.g. `Repeat(10)` could be an option but given that there is no clear preferred naming that is consistently used across languages we would basically force people to memorize the name, which seems rather suboptimal. 

We did not come to a clear conclusion and path forward despite our best efforts. 
One path we haven't explored that might give some inspiration is to think about whether some limited form of a "yielding loop" expression (similar to `[for i in 1..10 do yield i]` in F#) could be interesting and sufficiently compact without being overly cryptic.

Reply:
After thinking some more: 
While I think having something similar like Python's list comprehension might be valuable to avoid the awkwardness if the same thing needs to be expressed with the current library support, any reasonable way of expressing that I can come with is not compact/convenient enough to be a replacement for the new array expressions. However, I think it might make sense to fully buy into and expand on the notion that "things within rectangular brackets are special", and allow for a dedicated syntax to instantiate arrays of a certain length populated with a given value. I am thinking that something like 
`["", length=10]` might be a good option.

Reply:
Honest question, but do we want custom syntax because we think it's more intuitive/easier to read/more discoverable than a function? Or do we want it because we can't come up with a naming convention for the function? :)

IMO, `Array.Constant` is a lot more intuitive and discoverable than any custom syntax could be, because it's plausibly guessable - start typing `Array.` and get completions. And, IMO, it's also more readable - no need to interpret custom syntax invented for a single scenario, but it's simply a normal function call that is used throughout the language.

Reply:
Following the summary I gave above my suggestion, having Array.Repeated and then likely at some point in the future e.g. List.Repeated is suboptimal because we are now forced to implement a separate Repeated function for each type rather than having the option to express it in terms of supporting certain kinds of types. It is also awkward to extend to the multidimensional case and requires that we default import Array. `["", size=10]` on the other hand is easy enough to read, extensible to the multidimensional case, and more or less aligned with other syntax sugar we should consider adding for building arrays and similar types in the future.

Reply:
I am not sure why this applies to `Array.Repeated`, but not the rest of the `Microsoft.Quantum.Arrays` namespace. That entire namespace may be deprecated and generalized if/when Q# supports type classes. Why does `ConstantArray` need its own syntax, but not any of the rest of the functions? (Besides `EmptyArray`, but syntax for that naturally falls out of the general array literal syntax.)

Reply:
I am not sure I follow? My logic is that indeed that a good part of the M.Q.Arrays namespace won't be specific to arrays and in time hence migrate to a M.Q.Collection or M.Q.Enumerable or similar namespace. Having an Collection.Repeated however doesn't make sense, since it is not clear what Repeated needs to construct given only its argument. The only option hence is to either have an awkward split where functions like `Mapped` etc are in the collection namespace whereas `Repeated` is duplicated in the namespace/module for each type, or have all functions separate for each type (Array.Mapped, List.Mapped, etc), which adds unnecessary verbosity since in these cases the type can be determined by the argument.

Reply:
> Having an Collection.Repeated however doesn't make sense, since it is not clear what Repeated needs to construct given only its argument.

I don't understand this. We are already introducing return type inference to support `[]`. That would mean `Collection.Repeated` could infer the type based on the context in which the return value is used.

Reply:
Also, there are more collection types than we could reasonably invent syntax for. While arrays have literals, lists may or may not have literals in the future, and there are also lazy sequences, dictionaries, etc. Those types will probably need there own `Seq.Repeated`, `Seq.Empty`, `Dict.Empty`, etc. I think it's reasonable to have a type-specific namespace/module for things that weren't generalized to a type class.

Reply:
I believe the syntax for the relevant types would be possible to keep pretty much the same. Right on the inferring the type argument based on the usage for Repeat. That would then lead to using Repeat on its own, which brings us back to "it's something people will just need to learn by heart" without a good option to make it easily discoverable.

Reply:
> I believe the syntax for the relevant types would be possible to keep pretty much the same.

For all possible collection types? Even with type classes, it's not feasible to generalize everything. Type-specific modules is a common pattern even in languages that support type classes. I think it's a naming issue we'll need to confront eventually (`open ... as ...` is only sort of an ad-hoc way to do this).

> That would then lead to using Repeat on its own, which brings us back to "it's something people will just need to learn by heart" without a good option to make it easily discoverable.

If `Repeated` is in `Microsoft.Quantum.Core` or another auto-opened namespace, then I think it's still more discoverable than custom syntax due to code completion.

Reply:
Having specific namespaces for everything that is specific to a certain type makes perfect sense - I am not arguing there. What I think we should avoid is duplicating things that can be expressed more generally in the namespace for each type. 
I believe we both agree that we are down to the option of either having array creation be expressed simply as Repeat(10, "") with Repeat being in an automatically opened namespace or having it be custom syntax. Agreed?

Reply:
I'm not sure that `Array.Repeat`, `List.Repeat`, `Seq.Repeat`, etc. should be ruled out yet. I would need some more details about how exactly it should be generalized into a type class. Assuming we have some collection-ish type class, there are two options:

1. `Repeat` is not part of the type class definition. It can be implemented in terms of other collection functions.
2. `Repeat` is part of the type class definition. Every collection must provide an implementation of it.

I could be wrong, but I don't think that 1 is possible. That leaves 2, which means every type that has an instance of the collection type class must implement `Repeat` separately. This doesn't necessarily make sense, depending on how we define our type classes.

For example, it could be reasonable to say that a dictionary is a type of collection. But what does it mean to create a dictionary by repetition? The keys can't be repeated, but if a value is repeated, what key is assigned to each instance of the repeated value?

There may be other ways to organize the type classes that avoids this problem. But I'm not sure it's a given that `Repeat` can or should be generalized.

Reply:
Sets are another example. In the case of sets, it really doesn't make sense to construct a set by repetition. But they can be treated as a collection in most other ways (mapping, folding, etc.).

Reply:
Let's just look at what types Repeat should exist for (assuming the existence of all the things that we don't yet have). Repeat should exist for array, list, and sequence. As you point out, there is no meaningful way to define something like "initializer this collection with n times the same value" for sets and dictionaries, and similarly they also don't support e.g. numbers item access. I am not sure why considering set and dictionary to be of the same "kind" of type should be a given? 
You may agree that there is more than one type for which repeat should exist (at least for every type of arbitrary length that supports numbered item access I believe repeat makes sense). There should be a way to express one function called Repeat that can be used with all of those types. If that is not possible to express with type classes, then type classes don't fit the need we have.

Reply:
> There should be a way to express a function called Repeat that can be used with all of those types opposed to having to distinguish the function for each type by name. If that is not possible to express with type classes, then type classes don't fit the need we have.

Type classes are capable of doing this. At worst, you could define a RepeatableCollection for collection types where a `Repeat` could make sense...

But why is this a need? I think this assumption should be examined. It is not necessarily the case that everything in common needs to be generalized. What is the concrete need to write functions that can construct a collection by repetition, that are also polymorphic over the specific collection type?

Reply:
Indeed. Suppose that we want to define Repeat in the array namespace. This brings us back to that we have two options: Either make a fairly arbitrary split regarding which functions are in the type specific namespace and which ones are in a more general namespace, or put all functions which are similar to Repeat (e.g. Mapped) into the type specific namespace. 
You are advocating for the latter I believe, and I have some concerns based on my F# usage that does just that. I expect functions such as Mapped to be used fairly often. If they are defined for each type individually, then changing the type e.g. from list to sequence requires updating the code base. In addition to that hassle, having to qualify the name makes it more verbose (writing Array.Mapped instead of merely Mapped).

Reply:
>  Either make a fairly arbitrary split regarding which functions are in the type specific namespace and which ones are in a more general namespace, or put all functions which are similar to Repeat (e.g. Mapped) into the type specific namespace.
You are advocating for the latter I believe

No, I'm not suggesting we duplicate `Array.Mapped`, `List.Mapped`, etc. like F# does. I think `Mapped` should be generalized into a type class.

I'm saying that I don't think having a split between type class functionality and type-specific functionality is something that we should avoid at all costs. In most cases, I think it's OK to require specifying the exact type you want to construct - we already have to do this with array literals like `[1, 2, 3]`, and I don't see a push to generalize that into `MakeIndexedCollection(1, 2, 3)` that can be return-type-inferred into an array. And the syntax you're proposing, `["", size=10]`, already specifies the type to be an array, so it's no worse than `Array.Repeated` in that regard.

Reply:
The syntax you proposed implies some additional possibilities. Can you clarify if you want to support any of them?

* `let n = 4; ["", size = n]`
* `[size = 4, ""]`
* `["foo", "bar", size = 4]` -> `["foo", "bar", "foo", "bar"]` or error?
* `["foo", "bar", size = 3]` -> `["foo", "bar", "foo"]` or error?

Also, I assume you picked `size` instead of `length` to generalize to multidimensional arrays, but this unfortunately means that `Length(["", size = 4]) == 4`, but not `Size(["", size = 4]) == 4` unless we also add a `Size` function.

Reply:
Yes, I think it is a fair point to basically associate initialization functions with the type, and all transformation functions with the class of types, and I agree that the syntax solution also associates it with a type. 
I would only support the first one, not anything else.

Reply:
> associate initialization functions with the type, and all transformation functions with the class of types

Unfortunately, I don't think we can cleanly divide it like this, either - see for example `mempty` and `return`, both of which are constructors that are part of a type class:

```haskell
mempty :: Monoid a => a
return :: Monad m => a -> m a
```

My point is mostly that I don't think we need to have a 100% consistent rule here. I think when designing type classes in the standard library, we'll need to make decisions about what's in a type class and what's not, and not everything that can be generalized to more than one type should automatically go into a type class, necessarily. Even if `Repeated` is put into a type class, it's possible that both `SomeCollectionClass.Repeated` and `Array.Repeated` could work, where `Array.Repeated` is just the specific implementation for arrays.

Reply:
To summarize a bit of discussion that @bettinaheim and I had during our sync, I think we were able to come up with a few more ideas that may play nicely with future syntax (e.g.: array comprehensions and multidimensional arrays):

- `["" rep 10]`
- `["" dim 10]`
- `[10 of ""]`
- `["" | size=10]`

I'm personally a bit partial to `[10 of ""]` since it reads closer to English prose ("I want 10 of that"), but all of these generalize nicely to the 𝑛-d case (e.g.: `["" dim (3, 4)] == #[["", "", "", ""], ["", "", "", ""], ["", "", "", ""]]`), and can be made to work nicely with array comprehension. In particular, if we go with the size / shape on the left, then we could have array comprehensions like `[for i in 1..10 -> i * i]` (or `[for i in 1..10 => DrawRandomDouble(0. PI())]` for operations), while if we go with the element being repeated on the left, we could have comprehensions like `[i * i for i in 0..10]` (Python- or Julia-style).

Reply:
My slight aversion to `[10 of ""]` (which may just be a personal thing that most people don't perceive that way) is that when I am reading it, it I stumble upon it because starting with `[10` I am already thinking it is an array of integers with an element 10, and only when continuing I need to mentally revise to "no, it is actually an array of strings". That may be silly especially for a tiny piece of code, but suppose both expressions are larger and the expression is spread across multiple lines, it would require a double take for me to mentally parse it. I'm sure I wouldn't even notice that double take after a reading a few of these constructs, but I am wondering if it may be relevant in terms of how natural/easy it feels to use it.

Reply:
> That may be silly especially for a tiny piece of code, but suppose both expressions are larger and the expression is spread across multiple lines, it would require a double take for me to mentally parse it. I'm sure I wouldn't even notice that double take after a reading a few of these constructs, but I am wondering if it may be relevant in terms of how natural/easy it feels to use it.

I don't think it's silly at all; I'll admit I have a slight preference towards `of` since I read more it as prose, but I definitely take the point that it could be hard to read as it gets more complicated. Something like `[(nPlanes, nRows, nCols) of Complex(0.0, 0.0)]` to initialize a three-dimensional array of complex numbers could be a bit hard to read.

Reply:
I think that `let a = [1 dim (3, 4)];` is compelling. Will this be supported: `let x = 1; let d = 10; let a = [x dim d];`? How will it look like to initialize an array of qubits?

Reply:
Yes, that will be supported. 
If `q` is an initialized qubit then `[q dim 5]` would create an array with each item set to that qubit (though I would doubt that such an array is very useful). To get an initialized qubit, you have to use an allocation such as `use qs = Qubit[5];` (though we might want to revise that syntax slightly), just like it is the case currently.

Reply:
> To summarize a bit of discussion that @bettinaheim and I had during our sync, I think we were able to come up with a few more ideas that may play nicely with future syntax (e.g.: array comprehensions and multidimensional arrays):
> 
> * `["" rep 10]`
> * `["" dim 10]`
> * `[10 of ""]`
> * `["" | size=10]`
> 
> I'm personally a bit partial to `[10 of ""]` since it reads closer to English prose ("I want 10 of that"), but all of these generalize nicely to the 𝑛-d case (e.g.: `["" dim (3, 4)] == #[["", "", "", ""], ["", "", "", ""], ["", "", "", ""]]`), and can be made to work nicely with array comprehension. In particular, if we go with the size / shape on the left, then we could have array comprehensions like `[for i in 1..10 -> i * i]` (or `[for i in 1..10 => DrawRandomDouble(0. PI())]` for operations), while if we go with the element being repeated on the left, we could have comprehensions like `[i * i for i in 0..10]` (Python- or Julia-style).

I think from this list either  `["" | size=10]` or `[10 of ""]` are the most clear to me. 
What about `["" size 10]`? Then you get the clear keyword `size` and minimal additional syntax?

Reply:
One additional suggestion buried in the thread is also `["", size=10]`. Any thoughts on using `size` also in the multidimensional case? I would want to work is also for that.

Reply:
I like that `size` is a word (`rep` and `dim` are very cryptic to me), but I don't like that it's inconsistent with both `Length` (existing 1D arrays) and `Shape` (from the multidimensional arrays proposal).

Reply:
If we go with size here then we should choose Size for the multidimensional arrays as well. @cgranade

Reply:
+1 for keywords that are searchable. While I like the readability of `[10 of “”]` it does make it harder to find online or in code via search.

Reply:
With respect to the 𝑛-d proposal, the term "shape" aligns with Python / NumPy, Rust / ndarray, C# / NumSharp, and C++ / xtensor, but the term "size" is used by Julia, by C++ / Eigen. Meanwhile, the term "length" is used by C# (e.g.: `GetLength(int axis)`). Not content to pick a single term, MATLAB uses a mix of "size" and "shape," so no inspiration there.

While "shape" is more popular, I don't think it's so overwhelmingly so as to preclude moving to "size."

With respect to the immediate syntax question, though, can definitely get behind `["" of size 10]`, `["" of size (3, 4)]` and so forth (still sad about `of`, but I'll get over it, heh). I do want to quickly raise one other possibility inspired by Ruby, though. In particular, if we think of there being a parallel between array comprehension syntax and loop syntax, then `[expr for iter in arrayOrRange]` makes sense as the correspondance to `for (iter in arrayOrRange) { ... } `. We could imagine a new shorthand for `for` loops where we don't care about the iteration variable, following Ruby's `3.times do ... end` syntax:

```
times (10) {
     // this block repeats 10 times
}
```

In the same sense that this loop is short for `for (_ in 1..10) { ... }`, we could have a shorthand array comprehension `[expr times 10]`. The same argument would also hold for `[expr repeat 10]` being a shorthand for `repeat (10) { ... }` if we go that route. As @bettinaheim pointed out offline, this does have the disadvantage of mapping a bit less nicely onto the multidimensional case unless we have a nice syntax for multidimensional comprehensions (could be useful anyway).

Reply:
I like the idea of tying the syntax to loops, as it gives a commonality and makes the array generation seem seem like less of outlier. On the topic of shorthand, the `times` and `repeat` with an integer lines up with what I've been imagining for a proposal regarding integer limits on repeat-until-success loops. Then you could write something like
```qsharp
repeat (10) {
    ApplyOp(q);
}
until (M(q) == Zero);
```
And that would run the body either 10 times or until the condition is true, whichever comes first. But that's a separate conversation for another thread...

Reply:
`[Measure([PauliZ repeat 3], register) repeat until M(Head(register)) == One]`

Reply:
I was thinking about how that looks like together with comprehension and the multidimensional case. I am a bit worried that that if we have `<expr> size <expr>` it is harder to read and looks like a single operator expression. I think having an equal sign actually helps readability.
Here is what I've got, assuming syntax for multidimensional arrays and array comprehension: 
```qsharp
// all of these create an array of arrays of size 3
[0 size=(3,2)] 
[[x*x for x in 1..2] size=3] 
[[0 size=2], [0 size=3], [0 size=4]]
[[1,2], [3,4,5], [6,7,8,9]]

// all of these create a 2D array of size (3,2)
#[0 size=(3,2)] 
#[[x*x for x in 1..2] size=3] 
#[[0 size=2], [1 size=2], [2 size=2]]
#[[1,2], [3,4], [5,6]]

// all of these create a 3D array of size (4,3,2)
#[0 size=(4,3,2)] 
#[#[[x*y for x in 1..2] for y in 1..3] size=4] 
#[#[[1,2], [3,4], [5,6]], #[[1,2], [3,4], [5,6]], #[[1,2], [3,4], [5,6]], #[[1,2], [3,4], [5,6]]] 
```

Reply:
> I was thinking about how that looks like together with comprehension and the multidimensional case. I am a bit worried that that if we have `<expr> size <expr>` it is harder to read and looks like a single operator expression. I think having an equal sign actually helps readability.

Would we similarly want an `=` sign in array comprehensions, then, or some other operator to make it clear that comprehensions aren't expressions in their own right?

Reply:
Well, comprehensions are more expressions on their own right - they are like ranges in a sense. For to me implies an iteration, size does not.

Reply:
Alright, I think we need to come to a decision, and while I believe it is a good choice to keep the syntax restricted to within rectangular brackets for design reasons, there isn't an objectively best choice for the precise syntax. I am coming back to `[M(q), size=10]` for the following reasons: 
- I see it as a call to a constructor, and the two items within rectangular brackets as arguments. The distinction that it is a call is important since `M(q)` would be evaluated exactly once (also when the size is 0). This is in contrast to `[M(q) for _ in 1..10]` where `M(q)` would indeed be evaluated 10 times. 
- `size` is a full, unique, and short word that can be searched for in our docs.

Reply:
Thinking one step further, I believe it may actually be reasonable to support things like 
```qsharp
let buildTuple = (1,2,_); // callable to construct a tuple; the exact type of the callable will be inferred based on the first usage
let populateArray = [_, size=10]; // same here
```
in the future (though that should be worked out separately).

Reply:
Would it be possible to do it the other way, to have syntax that parallels `ConstantArray(_, "foo")` to partially apply the element but not the size? Something like `let bySize = ["foo", size=_]`?

Reply:
Yes, I would expect so.